{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "binding-medium",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "assisted-stranger",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['HG00379', 'HG00155']\n"
     ]
    }
   ],
   "source": [
    "import random\n",
    "import subprocess\n",
    "import sys\n",
    "\n",
    "# available samples\n",
    "chr1_file = '1000Genomes/ALL.chr1.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz'\n",
    "ls_samples = subprocess.check_output(\"bcftools query -l \" + chr1_file, shell=True).decode(sys.stdout.encoding)\n",
    "samples = ls_samples.split('\\n')\n",
    "\n",
    "# TODO: determine which are male/female (no X or Y chr files)\n",
    "parents = random.sample(samples, 2)\n",
    "print(parents)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "artistic-skiing",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bcftools view -s HG00379,HG00155 1000Genomes/ALL.chr1.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz | awk '{print $1\"\\t\"$2\"\\t\"$3\"\\t\"$4\"\\t\"$5\"\\t\"$10\"\\t\"$11\"\\n\"}' > chr1.txt\n"
     ]
    }
   ],
   "source": [
    "def sim_recomb(parent_ids, num_offspring):\n",
    "    for chrom in range(1, 23):\n",
    "        chrom_file = '1000Genomes/ALL.chr' + str(chrom) + '.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz'\n",
    "        queried_file = 'chr' + str(chrom) + '_parent_data.txt'\n",
    "        # query_cmd = subprocess.check_output('bcftools query -f [\\'%CHROM\\t%POS\\t%ID\\t%REF\\t%ALT\\t%GT\\n\\'] -s ' + ','.join(parent_ids) + ' ' + chrom_file + ' > chr' + str(chrom) + '.txt', shell=True) #.decode(sys.stdout.encoding)\n",
    "        # bcftools view -s HG00379,HG00155 1000Genomes/ALL.chr1.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz | awk '{print $1\"\\t\"$2\"\\t\"$3\"\\t\"$4\"\\t\"$5\"\\t\"$10\"\\t\"$11\"\\n\"}'\n",
    "        # bcftools query -f ['%CHROM\\t%POS\\t%ID\\t%REF\\t%ALT\\t%GT\\n'] -s HG00379,HG00155 1000Genomes/ALL.chr1.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz | awk  '{ if(NR%2 == 1) p = $6; else if (NR%2 == 0) print $0\"\\t\"p; }' | awk 'BEGIN{print \"CHROM\\tPOS\\tID\\tREF\\tALT\\tPARENT1\\tPARENT2\"}1'\n",
    "        bcf_cmd = 'bcftools query -f [\\'%CHROM\\t%POS\\t%ID\\t%REF\\t%ALT\\t%GT\\n\\'] -s ' + ','.join(parent_ids) + ' ' + chrom_file + ' | awk  \\'{ if(NR%2 == 1) p = $6; else if (NR%2 == 0) print $0\"\\\\t\"p; }\\' | awk \\'BEGIN{print \"CHROM\\tPOS\\tID\\tREF\\tALT\\t' + parent_id[0] + '\\\\t' + parent_id[1] + '\"}1\\''\n",
    "        min_pos = 'bcftools view -H ' + chrom_file + ' | awk \\'(NR==1)'\n",
    "        for o in range(num_offspring):\n",
    "            # 0 - orig chromatid1, 1 - recomb chromatid1, 2 - orig chromatid2, 3 - recomb chromatid2\n",
    "            chosen_chromatid = random.randint(0, 3)\n",
    "            if chosen_chromatid % 2 == 0: # non-recombined\n",
    "\n",
    "            else: # recombined\n",
    "                # choose crossover point\n",
    "                # find min and max pos values in file\n",
    "                # randomly choose value in range\n",
    "sim_recomb(parents, 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "informational-defensive",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3146: DtypeWarning: Columns (0,1) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "chr1_df = pd.read_csv('chr1.txt', sep='\\t', names=['CHROM', 'POS', 'ID', 'REF', 'ALT', parents[0], parents[1]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "empirical-elements",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>HG00379</th>\n",
       "      <th>HG00155</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>##fileformat=VCFv4.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>##FILTER=&lt;ID=PASS,Description=\"All</td>\n",
       "      <td>filters</td>\n",
       "      <td>passed\"&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>##fileDate=20150218</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>##reference=ftp://ftp.1000genomes.ebi.ac.uk//v...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>##source=1000GenomesPhase3Pipeline</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6468344</th>\n",
       "      <td>1</td>\n",
       "      <td>249240217</td>\n",
       "      <td>rs570135855;rs556759920</td>\n",
       "      <td>G</td>\n",
       "      <td>GT,T</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6468345</th>\n",
       "      <td>1</td>\n",
       "      <td>249240219</td>\n",
       "      <td>rs370070114</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>1|0</td>\n",
       "      <td>0|0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6468346</th>\n",
       "      <td>1</td>\n",
       "      <td>249240537</td>\n",
       "      <td>rs549505823</td>\n",
       "      <td>GGT</td>\n",
       "      <td>G</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6468347</th>\n",
       "      <td>1</td>\n",
       "      <td>249240539</td>\n",
       "      <td>rs545682552</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0|1</td>\n",
       "      <td>0|1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6468348</th>\n",
       "      <td>1</td>\n",
       "      <td>249240543</td>\n",
       "      <td>rs536805817</td>\n",
       "      <td>AGG</td>\n",
       "      <td>A</td>\n",
       "      <td>1|1</td>\n",
       "      <td>0|0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6468349 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                     CHROM        POS  \\\n",
       "0                                     ##fileformat=VCFv4.1        NaN   \n",
       "1                       ##FILTER=<ID=PASS,Description=\"All    filters   \n",
       "2                                      ##fileDate=20150218        NaN   \n",
       "3        ##reference=ftp://ftp.1000genomes.ebi.ac.uk//v...        NaN   \n",
       "4                       ##source=1000GenomesPhase3Pipeline        NaN   \n",
       "...                                                    ...        ...   \n",
       "6468344                                                  1  249240217   \n",
       "6468345                                                  1  249240219   \n",
       "6468346                                                  1  249240537   \n",
       "6468347                                                  1  249240539   \n",
       "6468348                                                  1  249240543   \n",
       "\n",
       "                              ID  REF   ALT HG00379 HG00155  \n",
       "0                            NaN  NaN   NaN     NaN     NaN  \n",
       "1                       passed\">  NaN   NaN     NaN     NaN  \n",
       "2                            NaN  NaN   NaN     NaN     NaN  \n",
       "3                            NaN  NaN   NaN     NaN     NaN  \n",
       "4                            NaN  NaN   NaN     NaN     NaN  \n",
       "...                          ...  ...   ...     ...     ...  \n",
       "6468344  rs570135855;rs556759920    G  GT,T     0|0     0|0  \n",
       "6468345              rs370070114    A     T     1|0     0|0  \n",
       "6468346              rs549505823  GGT     G     0|0     0|0  \n",
       "6468347              rs545682552    T     G     0|1     0|1  \n",
       "6468348              rs536805817  AGG     A     1|1     0|0  \n",
       "\n",
       "[6468349 rows x 7 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr1_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "pursuant-recommendation",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
